gno.land/pkg/sdk/vm/builtins.go

```diff
@@ -62,19 +65,59 @@ func (bnk *SDKBanker) RemoveCoin(b32addr crypto.Bech32Address, denom string, amo
 type SDKParams struct {
 	vmk *VMKeeper
 	ctx sdk.Context
+	// The curRealmPath is used to track the current realm accessing the SDKParams from the VM.
+	// It serves as a safeguard to control access from the VM.
+	curRealmPath string
 }
 
+// These are the native function implementations bound with standard libraries in Gno.
+// All methods of this struct are not supposed to be called from outside vm/stdlibs/std.
 func NewSDKParams(vmk *VMKeeper, ctx sdk.Context) *SDKParams {
 	return &SDKParams{
 		vmk: vmk,
 		ctx: ctx,
 	}
 }
 
-func (prm *SDKParams) SetString(key, value string)      { prm.vmk.prmk.SetString(prm.ctx, key, value) }
-func (prm *SDKParams) SetBool(key string, value bool)   { prm.vmk.prmk.SetBool(prm.ctx, key, value) }
-func (prm *SDKParams) SetInt64(key string, value int64) { prm.vmk.prmk.SetInt64(prm.ctx, key, value) }
+func (prm *SDKParams) SetString(key, value string) {
+	prm.assertRealmAccess()
+	prm.vmk.prmk.SetString(prm.ctx, key, value)
+}
+
+func (prm *SDKParams) SetBool(key string, value bool) {
+	prm.assertRealmAccess()
+	realmParamKey := fmt.Sprintf("%s.%s", prm.curRealmPath, lockSendKey)
+	if key == realmParamKey {
```

Comment:
this can't be the right design. 

1. curRealmPath itself is weird; it's a variable field of SDKParams, presumably it can change values, even though it's only supposed to be r/sys/params today, it would be easy to suggest a change that removes this restriction. then, this `if key == realmParamKey` condition could succeed where it should not, because another curRealmPath tried to use lockSendKey.

2. we're going to end up with all these custom prm.vmk.xxx.Yyy() injections in this file, but they would be grouped by type. I'm not sure exactly how, but really they should be grouped by the realm part of the key, if anything. And adding onto the point 1 above, the realm path should be extracted from the key, and compared against a const, not depend on curRealmPath. 

thinking...

Reply:
Agree,  see https://github.com/gnolang/gno/pull/3176#discussion_r1926137202

Reply:
done